contracts/global-extensions/GlobalOptimisticAuctionRebalanceExtension.sol

```diff
@@ -0,0 +1,328 @@
+
+        bytes32 assertionId = settings.optimisticParams.optimisticOracleV3.assertTruth(
+            claim,
+            msg.sender,
+            address(this),
+            address(0),
+            settings.optimisticParams.liveness,
+            settings.optimisticParams.collateral,
+            settings.optimisticParams.bondAmount,
+            settings.optimisticParams.identifier,
+            bytes32(0)
+        );
+
+        assertionIds[proposalHash] = assertionId;
+        proposalHashes[assertionId] = proposalHash;
+
+    }
+   
+    /**
+     * @dev OPERATOR ONLY: Checks that the old components array matches the current components array and then invokes the 
+     * AuctionRebalanceModuleV1 startRebalance function.
+     * 
+     * Refer to AuctionRebalanceModuleV1 for function specific restrictions.
+     *
+     * @param _quoteAsset                   ERC20 token used as the quote asset in auctions.
+     * @param _oldComponents                Addresses of existing components in the SetToken.
+     * @param _newComponents                Addresses of new components to be added.
+     * @param _newComponentsAuctionParams   AuctionExecutionParams for new components, indexed corresponding to _newComponents.
+     * @param _oldComponentsAuctionParams   AuctionExecutionParams for existing components, indexed corresponding to
+     *                                      the current component positions. Set to 0 for components being removed.
+     * @param _shouldLockSetToken           Indicates if the rebalance should lock the SetToken.
+     * @param _rebalanceDuration            Duration of the rebalance in seconds.
+     * @param _positionMultiplier           Position multiplier at the time target units were calculated.
+     */
+    function startRebalance(
+        ISetToken _setToken,
+        IERC20 _quoteAsset,
+        address[] memory _oldComponents,
+        address[] memory _newComponents,
+        AuctionExecutionParams[] memory _newComponentsAuctionParams,
+        AuctionExecutionParams[] memory _oldComponentsAuctionParams,
+        bool _shouldLockSetToken,
+        uint256 _rebalanceDuration,
+        uint256 _positionMultiplier
+    )
+        external
+        override
+    {
+        bytes32 proposalHash = keccak256(abi.encode(
+            _setToken,
+            _quoteAsset,
+            _oldComponents,
+            _newComponents,
+            _newComponentsAuctionParams,
+            _oldComponentsAuctionParams,
+            _shouldLockSetToken,
+            _rebalanceDuration,
+            _positionMultiplier
+        ));
+
+        bytes32 assertionId = assertionIds[proposalHash];
+        require(assertionId != bytes32(0), "Proposal hash does not exist");
+
+        delete assertionIds[proposalHash];
+        delete proposalHashes[assertionId];
+
+        ProductSettings memory settings = productSettings[_setToken];
+        settings.optimisticParams.optimisticOracleV3.settleAndGetAssertionResult(assertionId);
```

Comment:
this will revert if the assertion result is invalid? can we add some comments and note in the function natspec?

Reply:
Also what happens if someone else settled this assertion already by manually calling the oracle. Is this possible ? If so can this brick this method by letting this call revert ?

Reply:
Also why are we ignoring the assertion result here ? Shouldn't we check that the assertion passed ?

Reply:
should have test scenarios for all these

Reply:
@ckoopmann @pblivin0x 

https://github.com/UMAprotocol/protocol/blob/cef7d8880a2168eeb30f54151a8d40b274fbd381/packages/core/contracts/optimistic-governor/implementation/OptimisticGovernor.sol#L319-L321

This is heavily inspired by the Optimistic Governor contract. 

> this will revert if the assertion result is invalid?

If the assertion were invalid or disputed, the assertion [ above ](https://github.com/UMAprotocol/protocol/blob/cef7d8880a2168eeb30f54151a8d40b274fbd381/packages/core/contracts/optimistic-governor/implementation/OptimisticGovernor.sol#L313C8-L313C8) would have reverted the transaction as a callback would run on disputes to remove the assertionId from the mapping. 

> Also what happens if someone else settled this assertion already by manually calling the oracle. Is this possible ? If so can this brick this method by letting this call revert ?

This function would just get the result in that case and if it had been disputed, than this code would not be reached. Fetching the same result twice does not make this call revert. 

settleAndGetAssertionResult on the oo works as a settle if needed and get the result for and does not revert if called on an already settled assertionid. 

> Also why are we ignoring the assertion result here ? Shouldn't we check that the assertion passed ?

I explain this above and I agree that this reads rather unintuitively. Please take a look at my response, and if you need no further clarification, maybe suggest a comment or refactor that makes this more clear. I can think to maybe add an in-line comment on the assertion that mirrors the documentation on the [ OptimisticGovernor.sol](https://github.com/UMAprotocol/protocol/blob/cef7d8880a2168eeb30f54151a8d40b274fbd381/packages/core/contracts/optimistic-governor/implementation/OptimisticGovernor.sol#L309C9-L312C20)
```
        // This will reject the transaction if the proposal hash generated from the inputs does not have the associated
        // assertionId stored. This is possible when a) the transactions have not been proposed, b) transactions have
        // already been executed, c) the proposal was disputed or d) the proposal was deleted after Optimistic Oracle V3
        // upgrade.
```